flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+  }
+  // Unknown location for intrinsics.
+  return converter.genLocation();
+}
+
+Fortran::evaluate::characteristics::Procedure
+Fortran::lower::CallerInterface::characterize() const {
+  auto &foldingContext = converter.getFoldingContext();
+  auto characteristic =
+      Fortran::evaluate::characteristics::Procedure::Characterize(
+          procRef.proc(), foldingContext.intrinsics());
+  assert(characteristic && "Failed to get characteristic from procRef");
+  // The characteristic may not contain the argument characteristic if no
+  // the ProcedureDesignator has no interface.
+  if (!characteristic->HasExplicitInterface()) {
+    for (const auto &arg : procRef.arguments()) {
+      // Argument cannot be optional with implicit interface
+      const auto *expr = arg.value().UnwrapExpr();
+      assert(expr &&
+             "argument in call with implicit interface cannot be assumed type");
+      auto argCharacteristic =
+          Fortran::evaluate::characteristics::DummyArgument::FromActual(
+              "actual", *expr, foldingContext);
+      assert(argCharacteristic &&
+             "failed to characterize argument in implicit call");
+      characteristic->dummyArguments.emplace_back(
+          std::move(*argCharacteristic));
+    }
+  }
+  return *characteristic;
+}
+
+void Fortran::lower::CallerInterface::placeInput(
+    const PassedEntity &passedEntity, mlir::Value arg) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         passedEntity.firArgument >= 0 &&
+         passedEntity.passBy != CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = arg;
+}
+
+void Fortran::lower::CallerInterface::placeAddressAndLengthInput(
+    const PassedEntity &passedEntity, mlir::Value addr, mlir::Value len) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         static_cast<int>(actualInputs.size()) > passedEntity.firLength &&
+         passedEntity.firArgument >= 0 && passedEntity.firLength >= 0 &&
+         passedEntity.passBy == CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = addr;
+  actualInputs[passedEntity.firLength] = len;
+}
+
+bool Fortran::lower::CallerInterface::verifyActualInputs() const {
+  if (getNumFIRArguments() != actualInputs.size())
+    return false;
+  for (auto arg : actualInputs) {
+    if (!arg)
+      return false;
+  }
+  return true;
+}
+
+template <typename T>
+static inline auto AsGenericExpr(T e) {
+  return Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(e));
+}
+
+mlir::Value Fortran::lower::CallerInterface::getResultLength() {
+  // FIXME: technically, this is a specification expression,
+  // so it should be evaluated on entry of the region we are
+  // in, it can go wrong if the specification expression
+  // uses a symbol that may have change.
```

Comment:
Variable properties (including the special return variable) in the symbol map should be sorted and initialized in a legal canonical order for argument already. That would mean, say a LEN property, is established correctly at entry into the function. The variable associated with the LEN property then behaves like any other variable.

```
SUBROUTINE foo(x, n)
   CHARACTER(LEN=n) x
```

Then the symbol "x" is boxed with the ssa value of "n" on entry. The location of "n" itself can be updated freely thereafter.

The same should be done wrt the designated return variable.

Reply:
The case discussed in the comment is on the caller side, so it is the length of the function symbol that is not evaluated because its not a variable (On the calllee side, not problem, it is a variable).
Take the following example:
```
subroutine foo(n)
  integer :: n
  ! bar is a function so it is not instantiated in instantiateLocal and "n" is not evaluated here
  character(len=n) :: bar 
  n = n+1
  ! We would evaluate n here and that is wrong, since it was modified since function entry
  print *, bar()
end

function bar()
 character(*) :: bar
 bar = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
end function

 ! should print the nth first letters of the alphabet
 n = 10
 call foo(n)
end
```

That being said, gfortran gets it wrong also and prints 11 characters instead of 10 and pgfortran crashes at compile time, so.... kind of a corner case, and maybe forbidden in Fortran ? (I did not find the constraint). Hence, FIXME for now.